### ISSUE TITLE
figure out how FM will work.

### ISSUE THREAD

**Issue by thenfour** [thenfour/Clarinoid#84]
oscillators will need envelopes only for modulation use

-------

let's go for something a bit more workflow-oriented. I think fm support can be added with just 2 additional params:

* FM Algo - selecting which oscillators are carriers & which are modulators
* FM Feedback - controls a feedback modulation for each osc.

When an oscillator is a modulator, it is no longer routed out as an audible signal. It's automatically routed through a modulation based on the FM algo.

Otherwise everything is the same for the user.

Regarding the signal, it doesn't pass through the breath filter before routing back into the modulation matrix, so you get a clean constant signal which you can modulate with envelopes etc.

If this feels too awkward, it may be interesting to consider adding an amplitude envelope to modulators, but for a 1st POC it's not necessary.

**Comment by thenfour** [comment]
need to remember that the FM synths I'm used to (yamaha) are not actually freq modulation, they are phase modulation. I believe this accounts for the confusion / errors. It's also simpler in concept because frequency is in Hz, whereas phase is always 0-1 which corresponds well to an incoming modulator.

**Comment by thenfour** [comment]
at this point FM is working pretty well and the only problems left are regarding limitations of modulation. for example if you modulate amplitude, you are always multiplying by the modulator so the resulting wave :

1. can never be louder than the original (not a big deal, just a note)
2. can never be non-zero when the modulator is zero.

2 is important because for example if you want a baseline amplitude then you're out of luck. this feels like a common case.

Another problem with FM I'm finding is that it's so dependent on attacks. FM'ed waves add brightness, and this is mostly associated with attacks. Think FM slap bass, EPiano, marimbas... Sustaining FM instruments tend to have a very common sound. The variety of FM synthesis feels concentrated in the attack.

Basically, what's the role of breath in an FM instrument? Hard to say so far. I think this will need more experimenting.
